### PR TITLE
Add support for beta-NLL loss

### PR DESCRIPTION
**Pull request by kashif**
from the paper [On the Pitfalls of Heteroscedastic  Uncertainty Estimation with Probabilistic Neural Networks](https://openreview.net/pdf?id=aPOpXlnV1T) by Seitzer et al. 2022

official code repo: https://github.com/martius-lab/beta-nll

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

File: src/gluonts/torch/modules/loss.py [hunk 1]
```diff
@@ -48,10 +49,31 @@ def forward(
 
 
 class NegativeLogLikelihood(DistributionLoss):
+    def __init__(self, beta: float = 0.0):
+        """
```
Comment (lostella) [inline src/gluonts/torch/modules/loss.py:53]: Could you move this docstring to the class directly, instead of the constructor? So it’s consistent with the rest

File: src/gluonts/torch/modules/loss.py [hunk 2]
```diff
@@ -48,10 +49,31 @@ def forward(
...
+        self.beta = beta
+
     def forward(
         self, input: torch.distributions.Distribution, target: torch.Tensor
     ) -> torch.Tensor:
-        return -input.log_prob(target)
+        nll = -input.log_prob(target)
+        if self.beta > 0.0:
+            if type(input) == TransformedDistribution:
+                variance = input.base_dist.variance
```
Comment (lostella) [inline src/gluonts/torch/modules/loss.py:72]: Is this right?

File: src/gluonts/torch/modules/loss.py [hunk 3]
```diff
@@ -48,10 +49,31 @@ def forward(
...
 class NegativeLogLikelihood(DistributionLoss):
+    def __init__(self, beta: float = 0.0):
+        """
+        Compute the negative log likelihood loss.
+
+        Parameters
+        ----------
+        beta
+            beta parameter from the paper: "On the Pitfalls of Heteroscedastic
+            Uncertainty Estimation with Probabilistic Neural Networks" Seitzer et al. 2022
```
Comment (lostella) [inline src/gluonts/torch/modules/loss.py:60]: Flake8 complaining about too long lines (limit is 79 characters)

File: src/gluonts/torch/modules/loss.py [hunk 4]
```diff
@@ -48,10 +49,31 @@ def forward(
...
 
 class NegativeLogLikelihood(DistributionLoss):
+    def __init__(self, beta: float = 0.0):
+        """
+        Compute the negative log likelihood loss.
+
+        Parameters
+        ----------
+        beta
+            beta parameter from the paper: "On the Pitfalls of Heteroscedastic
```
Comment (lostella) [inline src/gluonts/torch/modules/loss.py:59]: Maybe you can suggest the range (0, 1) for this hyperparameter, and reasonable values: 0 recovers NLL, 0.5 appears to be a reasonable choice (at least from the results in the paper)

**Review by lostella — CHANGES_REQUESTED** [review]
Straight out of ICLR, thanks for this! I have left some comments

**Comment by kashif** [comment]
yes so I have to figure out how the method would work with TransformerDistribution with an affine transform coming from the scale... 

let me move this to draft for now...

**Comment by lostella** [comment]
> yes so I have to figure out how the method would work with TransformerDistribution with an affine transform coming from the scale...

Maybe it’s fine to support the simple case where only one affine transformation is there (multiple affine transformations should be collapsed into a single one anyway).

**Comment by kashif** [comment]
right... but the issue is i believe with the scale the variance needs to be scaled too... I am working through the maths... initial experiments are giving me nans for Gaussian and Student-t...

**Comment by lostella** [comment]
> right... but the issue is, i believe with the scale, the variance needs to be scaled too... 

Yes, it does. For MXNet distributions we have an explicit type for affinely-transformed distributions: https://github.com/awslabs/gluon-ts/blob/925d44c71b7ed85f0154afe22198a9cfc30713eb/src/gluonts/mx/distribution/transformed_distribution.py#L227 

This allows to avoid having to handle these (not so) special cases explicitly elsewhere

**Comment by kashif** [comment]
yup! that is what i did but getting nans... let me push it so you can have a look

File: src/gluonts/torch/modules/distribution_output.py [hunk 5]
```diff
@@ -141,11 +141,16 @@ def distribution(
...
         else:
+            scale = 1.0 if scale is None else scale
             transform = AffineTransform(
                 loc=0.0 if loc is None else loc,
-                scale=1.0 if scale is None else scale,
+                scale=scale,
             )
-            return TransformedDistribution(distr, [transform])
+            transformed_distr = TransformedDistribution(distr, [transform])
+            TransformedDistribution.variance = property(
```
Comment (lostella) [inline src/gluonts/torch/modules/distribution_output.py:150]: This seems a bit weird…

**Review by lostella — CHANGES_REQUESTED** [review]
How about something like:

```python
def variance(distr):
    if isinstance(distr, TransformedDistribution):
        # assert it’s transformed by single affine
        # return scaled variance
    return distr.variance
```

File: src/gluonts/torch/modules/distribution_output.py (hunk 5 above)
Comment (kashif) [reply src/gluonts/torch/modules/distribution_output.py]: you are right! I've fixed it I hope! please kindly review whenever you get time!

File: src/gluonts/torch/distributions/affine_transformed.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    AffineTransform,
+    Distribution,
+)
+
+
+class AffineTransformed(TransformedDistribution):
+    def __init__(self, base_distribution: Distribution, loc=None, scale=None):
+
+        self.scale = 1.0 if scale is None else scale
+        self.loc = 0.0 if loc is None else loc
```
Comment (lostella) [inline src/gluonts/torch/distributions/affine_transformed.py:25]:
```suggestion
    def __init__(self, base_distribution: Distribution, loc=0, scale=1):
        self.scale = scale
        self.loc = loc
```

File: src/gluonts/torch/modules/loss.py [hunk 7]
```diff
@@ -48,10 +48,29 @@ def forward(
...
+    """
+    Compute the negative log likelihood loss.
+
+    Parameters
+    ----------
+    beta: float in range (0, 1)
+        beta parameter from the paper: "On the Pitfalls of Heteroscedastic
+        Uncertainty Estimation with Probabilistic Neural Networks" by
+        Seitzer et al. 2022
+        https://openreview.net/forum?id=aPOpXlnV1T"""
```
Comment (lostella) [inline src/gluonts/torch/modules/loss.py:60]:
```suggestion
        https://openreview.net/forum?id=aPOpXlnV1T
    """
```

**Review by lostella — CHANGES_REQUESTED** [review]
Looks good to me, there are only a couple of minor things, plus what seems to be a circular dependency that is breaking the docs build: https://github.com/awslabs/gluon-ts/runs/6255414029?check_suite_focus=true

**Comment by kashif** [comment]
@lostella do you think the circular dependency is due to my changes? I cannot seem to figure out where its coming from...

**Comment by lostella** [comment]
> @lostella do you think the circular dependency is due to my changes? I cannot seem to figure out where its coming from...

I think so, I’ll try to sort it out

**Review by lostella — APPROVED** [review]
Looks good to me, thanks @kashif! 🚀

**Comment by kashif** [comment]
thanks @lostella 

My initial experiments were not too conclusive in the benefit of this scheme... my personal intuition would be to have the beta parameter decay down to zero or something small as training progresses... 

In any case, I would need more time to investigate this...
